### PR TITLE
[Modules] Quality Control Browser

### PR DESCRIPTION
**Pull request by llevitis**
I'm closing #3301 and opening this PR as trying to squash the 87 commits in that one was a nightmare. 

I've left the behavioral tab but can take it out as it currently is just the first tab from Data Team Helper. The filtering for 'Tarchive Status' is still a bit wonky and may not work for every project as the SeriesDescription for the same modalities differ between projects (CCNA & IBIS have different ones). For now this is a functional solution for at least IBIS and the Raisinbread dataset for t1, t2, and dti scans.

### CODE REVIEW

File: modules/quality_control/jsx/columnFormatterImg.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Modify behaviour of specified column cells in the Data Table component
+ * @param {string} column - column name
+ * @param {string} cell - cell content
+ * @param {arrray} rowData - array of cell contents for a specific row
```
Comment (MounaSafiHarab) [inline modules/quality_control/jsx/columnFormatterImg.js]: array (typo here and the line below)

Comment (driusan) [reply modules/quality_control/jsx/columnFormatterImg.js]: maybe it's a pirate array.

**Comment by MounaSafiHarab** [comment]
@llevitis 

1) I can not load the module (error_log is giving fatal error, probably one of the queries); we cna look at it together

2) it is unclear from the pull request where to access the module; I typed in the URL quality_control/ based on the php script name (I thought it would in the the Loris/Tools/ menu under Quality control but this takes us to data_team_helper)...
anyway adding the url manually does this "Loading..." and not loading anything due to the error in 1) above

screenshot from 2017-12-22 14-12-54

**Comment by llevitis** [comment]
@MounaSafiHarab As you suggested in person, one possible fix might be joining the tarchive_series & files table on SeriesUID & EchoTime. I will try implementing that change again. As for where the module appears within the menus, executing the patch should allow you to see it as Tools/Imaging QC. I will verify that the patch is working as intended.

image

**Comment by MounaSafiHarab** [comment]
> As for where the module appears within the menus, executing the patch should allow you to see it as Tools/Imaging QC. I will verify that the patch is working as intended.

@llevitis 
if there is a mysql patch, then I think you forgot to commit it (or am I not seeing it?)

**Comment by MounaSafiHarab** [comment]
@llevitis 

I still can not load it;
I get the module warning, and an ERROR with a missing file:

`
[Thu Jan 04 17:24:11.762223 2018] [:error] [pid 26720] [client 132.216.56.57:44040] Module quality_control is missing Module class descriptor. Modules missing a Module file will stop working in a futureversion of LORIS., referer: https://mouna-dev.loris.ca/quality_control/
[Thu Jan 04 17:24:14.292173 2018] [:error] [pid 26376] [client 132.216.56.57:44038] ERROR: File qualityControlIndex.js.map does not exist
`

File: modules/quality_control/php/NDB_Menu_Filter_quality_control.class.inc [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+                $query_params
+            );
+            if (substr("$scan_type", 0, 2) == "t1") {
+                $scans_done[$scan_type] = "T1_Scan_done";
+            } elseif (substr("$scan_type", 0, 3) == "t2w") {
+                $scans_done[$scan_type] = "T2_Scan_done";
+            } elseif (substr("$scan_type", 0, 3) == "dti") {
+                $scans_done[$scan_type] = "dti_Scan_done";
+            } else {
+                $scans_done[$scan_type] = $scan_type . "_Scan_done";
```
Comment (MounaSafiHarab) [inline modules/quality_control/php/NDB_Menu_Filter_quality_control.class.inc]: @llevitis 
I am not sure why it is necessary to hardcode t1, t2w and dti. This won't work across projects (for example, some projects might opt to call dti as dwi instead). Isn't the `else` part of those `if/elseif` statements the most generic way of doing what you are trying to accomplish?

File: modules/quality_control/php/NDB_Menu_Filter_quality_control_behavioral.class.inc [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+     * Override default behaviour, since the page is
+     * loaded from qualityControlIndex.js
+     *
+     * @return void
+     * @access public
+     */
+    function display()
+    {
+        return null;
+    }
```
Comment (driusan) [inline modules/quality_control/php/NDB_Menu_Filter_quality_control_behavioral.class.inc]: This shouldn't be here. This was a hack that Alex used once, which was replaced by the `$skipTemplate` class variable

File: modules/quality_control/jsx/columnFormatterImg.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Modify behaviour of specified column cells in the Data Table component
+ * @param {string} column - column name
+ * @param {string} cell - cell content
+ * @param {array} rowData - array of cell contents for a specific row
+ * @param {array} rowHeaders - array of table headers (column names)
+ * @return {*} a formated table cell for a given column
+ */
+function formatColumn(column, cell, rowData, rowHeaders) {
```
Comment (driusan) [inline modules/quality_control/jsx/columnFormatterImg.js]: I don't think this should be a global function, it should be a method on the class which is using it. The `formatColumn` function convention predates us converting most things to react, when there was no class to own the function, and doesn't make sense for new modules which are written entirely in React

File: modules/quality_control/jsx/columnFormatterImg.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  // If a column if set as hidden, don't display it
+  if (loris.hiddenHeaders.indexOf(column) > -1) {
```
Comment (driusan) [inline modules/quality_control/jsx/columnFormatterImg.js]: I don't see why this uses a global variable on the Loris object. If it's constant, it should just be inlined into the formatter, and if it's variable it should be a state on the React component that defines it which is passed around as a prop.

**Comment by sruthymathew123** [comment]
@llevitis 
QC module loaded on my VM
Is it just the imaging tab that I really want to test?

**Comment by kongtiaowang** [comment]
phpcs

**Comment by sruthymathew123** [comment]
@llevitis 
qc_concern

**Comment by llevitis** [comment]
@sruthymathew123 No, it's not an issue for the module itself. Rather, it's an issue that the module is displaying.  The module is designed to query for the presence of specific scan types for a given tarchiveID by comparing all the corresponding tarchive_series values with ranges & values specified in the mri_protocol table. A value of 'Missing' for 'Tarchive' means that there is no scan for that tarchive matching the echo time ranges, repetition time ranges, etc for a scan type of t1w in the mri_protocol table.

**Comment by sruthymathew123** [comment]
@llevitis @driusan 
Along with the normalization PRs #3528(the sql patch run is necessary to test this module),  #3523,
the module loaded fine on IBIS data set. Everything seems to be working.

@llevitis Quick question; populating all the inactive sites in the site filter is anything intentional? Ref: $siteList = Utility::getSiteList(false);

**Comment by llevitis** [comment]
@sruthymathew123 Good catch! That wasn't intentional, and I'm no longer passing "false" to getSiteList.

**Comment by sruthymathew123** [comment]
@llevitis @driusan I am adding "passed manual tests" label to this PR as this module seems to be working on IBIS data set.  Also adding a blocked label to this PR as this PR will break unless this #3528 is merged.

**Comment by ridz1208** [comment]
@sruthymathew123 @llevitis removed Passed manual test as changes are still expected to occur

File: SQL/Archive/19.1/2018-01-23-qc_module_patch.sql [hunk 6]
```diff
@@ -0,0 +1 @@
+INSERT INTO LorisMenu (Label, Link, Parent, OrderNumber) VALUES ('Quality Control','quality_control/', (SELECT ID FROM LorisMenu as L WHERE Label='Tools'), 10);
```
Comment (ridz1208) [inline SQL/Archive/19.1/2018-01-23-qc_module_patch.sql]: this is going to the wrong directory

**Comment by sruthymathew123** [comment]
#3973

**Comment by llevitis** [comment]
As discussed with @ridz1208 at yesterday's LORIS meeting, this PR is waiting for #3973 to be merged. Im adding the `needs_work` label because the code also needs to be changed to namespace conventions.

**Pull request closed without merging**